OTVision/detect/yolo.py

```diff
@@ -0,0 +1,491 @@
+from typing import Any, Union
+
+import torch
+from cv2 import CAP_PROP_FPS, VideoCapture
+
+from OTVision.config import CONFIG
+from OTVision.helpers.files import has_filetype
+from OTVision.helpers.log import log
+
+
+class NoVideoError(Exception):
+    pass
+
+
+class VideoFoundError(Exception):
+    pass
+
+
+class YOLOv5ModelNotFoundError(Exception):
+    pass
+
+
+def detect_video(
+    file: Path,
+    model: Union[torch.nn.Module, None] = None,
+    weights: str = CONFIG["DETECT"]["YOLO"]["WEIGHTS"],
+    conf: float = CONFIG["DETECT"]["YOLO"]["CONF"],
+    iou: float = CONFIG["DETECT"]["YOLO"]["IOU"],
+    size: int = CONFIG["DETECT"]["YOLO"]["IMGSIZE"],
+    chunksize: int = CONFIG["DETECT"]["YOLO"]["CHUNKSIZE"],
+    normalized: bool = CONFIG["DETECT"]["YOLO"]["NORMALIZED"],
+) -> dict[str, dict]:  # TODO: Type hint nested dict during refactoring
+    """Detect and classify bounding boxes in videos using YOLOv5
+    Args:
+        file (Path): files to detect.
+        model (torch.nn.Module): Yolo model to detect with.
+        weights (str, optional): Weigths, if no model passed. Defaults to "yolov5s".
+        conf (float, optional): Output confidence, if no model passed. Defaults to 0.25.
+        iou (float, optional): IOU param, if no model passed. Defaults to 0.45.
+        size (int, optional): Frame size for detection. Defaults to 640.
+        chunksize (int, optional): Number of files per detection chunk. Defaults to 0.
+        normalized (bool, optional): Coords in % of image/frame size (True) or pixels
+            (False). Defaults to False.
+    Returns:
+        dict[str, dict]: Dict with subdicts of metadata and actual detections
+    """
+    if model is None:
+        model = loadmodel(weights, conf, iou)
+
+    yolo_detections: list = []
+    t1 = perf_counter()
+
+    if not has_filetype(file, CONFIG["FILETYPES"]["VID"]):
+        raise NoVideoError(f"The file: {file} is not a video!")
+
+    cap = VideoCapture(str(file))
+    batch_no = 0
+
+    log.info(f"Run detection on video: {file}")
+
+    got_frame = True
+    t_loop_overhead = 0.0
+    while got_frame:
+        t_start = perf_counter()
+        got_frame, img_batch = _get_batch_of_frames(cap, chunksize)
+
+        t_get_batch = perf_counter()
+
+        if not img_batch:
+            break
+
+        # What purpose does this transformation have
+        transformed_batch = list(map(lambda frame: frame[:, :, ::-1], img_batch))
+
+        t_trans = perf_counter()
+
+        results = model(transformed_batch, size)
+
+        t_det = perf_counter()
+
+        _add_detection_results(yolo_detections, results, normalized)
+
+        t_list = perf_counter()
+
+        _log_batch_performances_stats(
+            batch_no,
+            t_start,
+            t_get_batch,
+            t_trans,
+            t_det,
+            t_list,
+            t_loop_overhead,
+            len(img_batch),
+        )
+        batch_no += 1
+
+        width = cap.get(3)  # float
```

Comment:
Why is this inside the loop?
The variables are used outside the loop and `cap` is defined outside the loop.